\documentclass{article}
%\VignetteIndexEntry{PACVr}
%\VignetteKeywords{PACVr}
%\VignettePackage{PACVr}
%\VignetteEncoding{UTF-8}


% PACKAGE DEFINITION
\usepackage{fancyvrb}
\usepackage{fullpage}
\usepackage{listings}
\usepackage{xcolor}
\usepackage{float}
\usepackage{hyperref}
\usepackage{graphicx}
\usepackage{tcolorbox}

% CUSTOM FUNCTIONS
\tcbuselibrary{skins,breakable}
\newenvironment{BGVerbatim}
 {\VerbatimEnvironment
  \begin{tcolorbox}[
    breakable,
    colback=orange!10,
    spartan
  ]%
  \begin{Verbatim}}
 {\end{Verbatim}\end{tcolorbox}}

\author{Michael Gruenstaeudl, Nils Jenke}
\title{Using PACVr}


\begin{document}
\SweaveOpts{concordance=TRUE}
\maketitle
\tableofcontents

\section{Introduction}
PACVr visualizes the coverage depth of a complete plastid genome as well as the equality of its inverted repeat regions in relation to the circular, quadripartite genome structure and the location of individual genes. This vignette provides instructions for generating the necessary input files and for executing the software from within the R interpreter via function \texttt{PACVr.complete()}, and invocation from the Unix command-line shell via script \texttt{PACVr\_Rscript.R}. This vignette also illustrates the operation of PACVr on an empirical dataset co-supplied with the R package.

\section{Creating BAM file}
For PACVr to compute and visualize coverage depth along a given plastid genome, users must provide textual information on the mapping of sequence reads on the input genome. This information is supplied via a BAM file. The following is a minimal example code for generating a BAM file using the read mapper BWA (\href{https://doi.org/10.1093/bioinformatics/btp324}{Li and Durbin 2009}).
\begin{BGVerbatim}[fontsize=\normalsize]
bwa -index -a is <reference.fasta>
bwa aln <reference.fasta> <forward_reads.fastq> > <forwards_reads.sai>
bwa aln <reference.fasta> <backward_reads.fastq> > <backward_reads.sai>
bwa mem -t 6 <reference.fasta> <forwards_reads.fastq> <backward_reads.fastq> \
    | samtools view -Sb - > <Sample_BackMapping_bwa.bam>
samtools sort <Sample_BackMapping_bwa.bam> -o <Sample_BackMapping_bwa.sorted.bam>
samtools index <Sample_BackMapping_bwa.sorted.bam>
\end{BGVerbatim}

\section{Installation}
Prior to running PACVr, several dependencies need to be installed. Below are some strategies that have been successful in installing PACVr and all its dependencies on Linux (Arch Linux 4.18, Debian 9.9, and Ubuntu 18.10) and MacOS (HighSierra 10.13.6 and Mojave 10.14.6), respectively.

    \subsection{On Linux}
    To install \href{https://doi.org/10.1093/bioinformatics/btx699}{mosdepth}, open a shell and type:
    \begin{BGVerbatim}[fontsize=\normalsize]
    system("conda install -y mosdepth")
    \end{BGVerbatim}
    
    To install all R dependencies, open R and type:
    \begin{BGVerbatim}[fontsize=\normalsize]    
    # CRAN packages
    if (!require("pacman"))
        install.packages("pacman")
    pacman::p_load("RCircos", "optparse", install=TRUE)
    # Bioconductor packages
    if (!requireNamespace("BiocManager", quietly=TRUE))
        install.packages("BiocManager")
    if (!requireNamespace("genbankr", quietly=TRUE))
        BiocManager::install("genbankr")
    \end{BGVerbatim}

    \subsection{On MacOS}
    To install \href{https://doi.org/10.1093/bioinformatics/btx699}{mosdepth}, do:
    \begin{itemize}
    \itemsep0em
    \item Install \href{https://brew.sh/}{brew}
    \item Install \href{https://nim-lang.org/}{nim} via brew: \texttt{brew install nim}
    \item Install \href{https://www.htslib.org/}{hts-lib} via brew: \texttt{brew install htslib}
    \item Obtain the source code of \href{https://doi.org/10.1093/bioinformatics/btx699}{mosdepth} from its \href{https://github.com/brentp/mosdepth/releases}{Github page}
    \item In the unzipped mosdepth source code directory, open a terminal and type: \texttt{nimble build} (confirm when prompted) and \texttt{nim c -d:release mosdepth.nim}
    \item When in the unzipped mosdepth source code directory, confirm the compilation of mosdepth by typing \texttt{./mosdepth -h}
    \end{itemize}
    
    To install all R dependencies, open R and follow the instructions as given for Linux (see above).

%\pagebreak

\section{Executing PACVr via R interpreter}
PACVr can be executed from within the R interpreter via function \texttt{PACVr.complete()}.

\begin{BGVerbatim}[fontsize=\normalsize]
library(PACVr)
## Specify input files
gbkFile <- system.file("extdata", "MH161174/MH161174.gb", package="PACVr")
bamFile <- system.file("extdata", "MH161174/MH161174_PlastomeReadsOnly.sorted.bam", 
                       package="PACVr")
## Specify output file
outFile <- paste(tempdir(), "/MH161174_AssemblyCoverage_viz.pdf", sep="")
## Run PACVr
PACVr.complete(gbk.file=gbkFile, bam.file=bamFile, windowSize=250, 
               mosdepthCmd='mosdepth', logScale=FALSE, threshold=0.5,
               syntenyLineType=3, relative=TRUE, textSize=0.5,
               delete=TRUE, output=outFile)
\end{BGVerbatim}

\section{Executing PACVr via Unix shell}
PACVr can also be executed from the Unix command-line shell via script \texttt{PACVr\_Rscript.R}.
  
\begin{BGVerbatim}[fontsize=\normalsize]
Rscript ./inst/extdata/PACVr_Rscript.R \
     -k ./inst/extdata/MH161174.gb \
     -b ./inst/extdata/MH161174_PlastomeReadsOnly.sorted.bam \
     -o ./inst/extdata/MH899017_AssemblyCoverage_viz.pdf
\end{BGVerbatim}
  
\begin{figure}[H]
\centering
    \includegraphics[width=0.9\textwidth]{MH161174_AssemblyCoverage_viz.pdf}
    \caption{File \texttt{MH161174\_AssemblyCoverage\_viz.pdf} as generated via \texttt{PACVr.complete()}}
\centering
\end{figure}

\pagebreak

%\section{Modifying parameters}
%  Depending on which system PACVr will be executed

%\section{More Information}

\iffalse
\section{sessionInfo}
\begin{Verbatim}
> sessionInfo()
R version 3.3.3 (2017-03-06)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 9 (stretch)

locale:
 [1] LC_CTYPE=de_DE.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=de_DE.UTF-8        LC_COLLATE=de_DE.UTF-8    
 [5] LC_MONETARY=de_DE.UTF-8    LC_MESSAGES=de_DE.UTF-8   
 [7] LC_PAPER=de_DE.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=de_DE.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets 
[6] methods   base     

loaded via a namespace (and not attached):
[1] tools_3.3.3
\end{Verbatim}
\fi

\end{document}
